be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/SocialMediaHandler.scala

```diff
@@ -36,50 +40,46 @@ class SocialMediaHandler(dbRef: => AskableActorRef) extends MessageHandler {
 
   private def generateSocialChannel(lao_id: Hash): Channel = Channel(Channel.ROOT_CHANNEL_PREFIX + lao_id + Channel.SOCIAL_MEDIA_CHIRPS_PREFIX)
 
+  // the following function encodes the data and broadcast it
+  private def broadcastData[T](rpcMessage: JsonRpcRequest, data: T, channel: Channel, broadcastChannel: Channel): GraphMessage = {
+    val encodedData: Base64Data = Base64Data.encode(data.toString)
+    val broadcast: Future[GraphMessage] = dbBroadcast(rpcMessage, channel, encodedData, broadcastChannel)
+    Await.result(broadcast, duration)
+  }
+
   def handleAddChirp(rpcMessage: JsonRpcRequest): GraphMessage = {
-    writeAndPropagate(rpcMessage) match {
-      case Left(_) =>
-        val channelChirp: Channel = rpcMessage.getParamsChannel
-        channelChirp.decodeChannelLaoId match {
-          case Some(lao_id) =>
-            val broadcastChannel: Channel = generateSocialChannel(lao_id)
-            rpcMessage.getParamsMessage match {
-              case Some(params) =>
-                // we can't get the message_id as a Base64Data, it is a Hash
-                val chirp_id: Hash = params.message_id
-                val timestamp: Timestamp = params.decodedData.get.asInstanceOf[AddChirp].timestamp
-                val notifyAddChirp: NotifyAddChirp = NotifyAddChirp(chirp_id, channelChirp, timestamp)
-                val broadcastData: Base64Data = Base64Data.encode(notifyAddChirp.toJson.toString)
-                Await.result(dbBroadcast(rpcMessage, rpcMessage.getParamsChannel, broadcastData, broadcastChannel), duration)
-              case None => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "Server failed to extract chirp id for the broadcast", rpcMessage.id))
-            }
-          case None => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "Server failed to extract LAO id for the broadcast", rpcMessage.id))
-        }
-      case error @ Right(_) => error
-      case _                => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswerDatabase, rpcMessage.id))
+    val ask =
+      for {
+        _ <- extractLaoChannel(rpcMessage, "Server failed to extract LAO id for the broadcast")
+        _ <- extractParameters[AddChirp](rpcMessage, "Server failed to extract chirp id for the broadcast")
+        _ <- dbActor ? DbActor.WriteAndPropagate(rpcMessage.getParamsChannel, rpcMessage.getParamsMessage.get)
+      } yield ()
+
+    Await.ready(ask, duration).value match {
+      case Some(Success(_)) =>
+        val (chirp_id, channelChirp, data, broadcastChannel) = parametersToBroadcast[AddChirp](rpcMessage)
+        val notifyAddChirp: NotifyAddChirp = NotifyAddChirp(chirp_id, channelChirp, data.timestamp)
+        broadcastData[NotifyAddChirp](rpcMessage, notifyAddChirp, channelChirp, broadcastChannel)
```

Comment:
I just implemented this function `broadcastData`, which encodes and broadcast the data. It should be more readable. What do you think?

Reply:
It's becoming more concise and the intention remains pretty clear, so that's good.
However, it does add a level of indirection and hides where the `Await.result` happens.

I'm wondering: the new `broadcastData` function doesn't seem to be doing much. Would it make sense to change `dbBroadcast` to become a generic (i.e., `dbBroadcast[T]`) and move the `Base64Data.encode` there ? 
Then we'd only have to keep the `Await.result`.

Reply:
Yes that is true that it only moves the `Await.result`, thus it is not really useful. I removed it. 

I took in account your remark and I found out that the function does not need to be generic. To move the `Base64Data.encode` into the `dbBroadcast` function, we simply need to take a `String` instead of a `Hash` as a parameter for the `broadcastData`. I think that it is the maximum generalization that we can have for this function (for now on).